pip/wheel.py

```diff
@@ -39,6 +44,46 @@
 logger = logging.getLogger(__name__)
 
 
+def _cache_for_filename(cache_dir, sdistfilename):
+    """Return a directory to store cached wheels in for sdistfilename.
+
+    Because there are M wheels for any one sdist, we provide a directory
+    to cache them in, and then consult that directory when looking up
+    cache hits.
+
+    :param cache_dir: The cache_dir being used by pip.
+    :param sdistfilename: The filename of the sdist for which this will cache
+        wheels.
+    """
+    return os.path.join(cache_dir, 'wheels', sdistfilename)
+
+
+def cached_wheel(cache_dir, link):
+    if not cache_dir:
+        return link
+    if not link:
+        return link
+    if link.is_wheel:
+        return link
+    root = _cache_for_filename(cache_dir, link.filename)
+    try:
+        candidates = os.listdir(root)
```

Comment:
We should probably sort this similarly to how ``PackageFinder`` does, to make sure that we prefer a more specific wheel over a more generic wheel.